inc/restrictedformcriteria.class.php

```diff
@@ -0,0 +1,249 @@
+      global $DB;
+
+      // Intersect the given form with the form for which the user is whitelisted
+      $data = $DB->request([
+         'COUNT' => 'count',
+         'FROM'  => PluginFormcreatorForm::getTable(),
+         'WHERE' => [
+            ['id' => static::getListCriteriaSubQuery()],
+            ['id' => $form->fields['id']],
+         ]
+      ]);
+
+      $row = $data->current();
+      return $row['count'] > 0;
+   }
+
+   public static function countItemsToImport($input) : int {
+      return 1;
+   }
+
+   /**
+   * Prepare input data for adding the form
+   *
+   * @param array $input data used to add the item
+   *
+   * @return array the modified $input array
+   */
+   public function prepareInputForAdd($input) {
+      // generate a unique id
+      if (!isset($input['uuid']) || empty($input['uuid'])) {
+         $input['uuid'] = plugin_formcreator_getUuid();
+      }
+      return $input;
+   }
+
+   public function deleteObsoleteItems(CommonDBTM $container, array $exclude) : bool {
+      $keepCriteria = [
+         self::$items_id_1 => $container->getID(),
+      ];
+      if (count($exclude) > 0) {
+         $keepCriteria[] = ['NOT' => ['id' => $exclude]];
+      }
+      return $this->deleteByCriteria($keepCriteria);
+   }
+
+   /**
+    * Import a form's visibility criteria into the db
+    * @see PluginFormcreatorForm::importJson
+    *
+    * @param  PluginFormcreatorLinker  $linker  id of the parent form
+    * @param  array                    $input
+    * @param  int                      $container_id
+    * @return integer|false the item ID or false on error
+    */
+   public static function import(
+      PluginFormcreatorLinker $linker,
+      $input = [],
+      $container_id = 0
+   ) {
+      $itemtype2 = static::$itemtype_2;
+      $item2_input_key = "_" . strtolower($itemtype2);
+
+      if (!isset($input['uuid']) && !isset($input['id'])) {
+         throw new ImportFailureException(
+            sprintf(
+               'UUID or ID is mandatory for %1$s',
+               static::getTypeName(1)
+            )
+         );
+      }
+
+      $form_fk = PluginFormcreatorForm::getForeignKeyField();
+      $input[$form_fk] = $container_id;
+      $item = new static();
+      // Find an existing form to update, only if an UUID is available
+      $item_id = false;
+      /** @var string $id_key key to use as ID (id or uuid) */
+      $id_key = 'id';
+      if (isset($input['uuid'])) {
+         // Try to find an existing item to update
+         $id_key = 'uuid';
+         $item_id = plugin_formcreator_getFromDBByField(
+            $item,
+            'uuid',
+            $input['uuid']
+         );
+      }
+
+      // Set the linked item
+      $item2 = new $itemtype2();
+      $form_fk  = PluginFormcreatorForm::getForeignKeyField();
+      if (!plugin_formcreator_getFromDBByField($item2, 'name', $input[$item2_input_key])) {
+         // Item not found, lets ignore this import
+         return true;
```

Comment:
docblock does not allows returning true. Only the ID of the imported item or false on failure.

If I check in target_actor.class.php, when trying to add an actor to a target, failing to find the linked item (group, user or supplier) throws an import exception. I think an error is necessary as silently fail and continue may lead to undetected import error and may impact a workflow if the form is incomplete.

Reply:
Done f1959675.